### PR TITLE
Allow customizing consult-line to go to start or end of match

### PR DESCRIPTION
**Pull request by oantolin**
This adds a variable `consult-goto-start-of-match` controlling whether `consult-line` leaves you at the beginning or at the end of the match, as discussed in issue #26.

### CODE REVIEW

**Comment by minad** [comment]
Is there a customization option for isearch btw? I haven't found one. I found this post, https://endlessparentheses.com/leave-the-cursor-at-start-of-match-after-isearch.html, but this is not good since the cursor only goes to the beginning when leaving the search.

**Comment by oantolin** [comment]
I believe there is no boolean variable for isearch. There is an isearch-mode-end-hook, though. I set that hook to the following function:

```elisp
(defun isearch-exit-at-start ()
  "Exit search at the beginning of the current match."
  (unless (or isearch-mode-end-hook-quit
              (bound-and-true-p isearch-suspended)
              (not isearch-forward)
              (not isearch-other-end))
    (goto-char isearch-other-end)))
```

I also bind this command to `<S-return>` in `isearch-mode-map`:

```elisp
(defun isearch-exit-at-end ()
  "Exit search at the end of the current match."
  (interactive)
  (let ((isearch-other-end (point)))
    (isearch-exit))
  (unless isearch-forward (goto-char isearch-other-end)))
```

**Comment by oantolin** [comment]
Oh, sorry, I hadn't followed the link. It must be where I got the idea in the first place!

**Comment by oantolin** [comment]
I use a slightly different condition though, I wonder why I use that and not Sylvain Rousseau's condition... I can't remember.

**Comment by minad** [comment]
@oantolin Ok, I think there is no way to implement this isearch thing differently, since during the search the cursor must be at the end of the match in order to be able to step to the next match. The isearch-mode-hook seems to be the best one can do.

Regarding this issue - Now that we are introducing a configuration option here, I consider over-engineering it - I would like to offer three options: 'start, 'start-match, 'end-match. If setting 'start, the line matching will be disabled completely.

**Comment by minad** [comment]
> Regarding this issue - Now that we are introducing a configuration option here, I consider over-engineering it - I would like to offer three options: 'start, 'start-match, 'end-match. If setting 'start, the line matching will be disabled completely.

I quickly write the patch for this.

**Comment by oantolin** [comment]
I'd suggest being more explicit in the names of the options: `start-of-line`, `start-of-match`, `end-of-match`. Maybe toss in `end-of-line` for symmetry?

**Comment by oantolin** [comment]
Also the variable should be renamed with more options. Maybe `consult-line-point-placement`?

**Comment by minad** [comment]
> I'd suggest being more explicit in the names of the options: start-of-line, start-of-match, end-of-match. Maybe toss in end-of-line for symmetry?

Yes, also thought about end-of-line. I would like it for symmetry, but there are two reasons why I omitted it: 1. It is useless, 2. the other commands all jump to the beginning of the line, e.g. consult-imenu. I don't want to adjust them such that they also offer the end-of-line setting. So basically consult-line is only augmented by start-of-match/end-of-match and otherwise consult-line can be made to behave like consult-imenu and friends.

**Comment by oantolin** [comment]
Ah, OK, that makes sense, no `end-of-line`, then.

**Comment by oantolin** [comment]
I know you offered to do it, but if you'd like I can make the changes.

**Comment by oantolin** [comment]
What should the default be, though? Maybe `start-of-line` to match `consult-imenu` and the others?

**Comment by minad** [comment]
Ok, please go forward with it! start-of-match should be the default, since it is better. Still having the start-of-line option is sufficient if people want to make the behavior of the commands consistent.

**Comment by oantolin** [comment]
So I have heard of commit squashing but never done it before? Do you want me to make these a single commit?

**Comment by minad** [comment]
Yes squashing is a generally a good idea for single features!

File: consult.el [hunk 1]
```diff
@@ -927,22 +928,26 @@ CAND is the currently selected candidate."
       (let ((start 0)
```
Comment (minad) [inline consult.el:928]: The start-of-line check should be performed right in the beginning, just return pos directly.

File: consult.el [hunk 2]
```diff
@@ -927,22 +928,26 @@ CAND is the currently selected candidate."
...
             (end (length cand)))
         ;; Find match end position, remove characters from line end until matching fails
-        (let ((step 16))
-          (while (> step 0)
-            (while (and (> (- end step) 0)
-                        (completion-all-completions input (list (substring cand 0 (- end step))) nil 0))
-              (setq end (- end step)))
-            (setq step (/ step 2))))
+        (unless (eq consult-line-point-placement 'start-of-line)
```
Comment (minad) [inline consult.el:931]: The end computation must be done in every case, since end is used later.

Comment (oantolin) [reply consult.el:931]: I currently do the end computation for `start-of-match` and for `end-of-match`. That should be enough, no? The third case, `start-of-line` doesn't need it.

File: consult.el (hunk 1 above)
Comment (oantolin) [reply consult.el:928]: Sure, I guess it could be done that way. I'll change it.

File: consult.el (hunk 2 above)
Comment (minad) [reply consult.el:931]: Right, I misread. Sorry.

Comment (oantolin) [reply consult.el]: Now that the start-of-line case is handled earlier, I removed this check, but I still think it was correct before. 😉

**Comment by oantolin** [comment]
OK, I made the changes you suggested and figured out how to squash into a single commit. (I think.)

**Comment by minad** [comment]
Thanks, looks good. I slightly adjusted the names such that they match the standard Emacs terminology.

**Pull request closed without merging**